### PR TITLE
First release of OpenStack examples

### PR DESCRIPTION
**Pull request by Pansanel**
Set of examples for setting up a OpenStack Icehouse Cloud Controller, a Network node and a Compute node

### CODE REVIEW

**Comment by Pansanel** [comment]
This set of examples requires https://github.com/quattor/configuration-modules-core/pull/423. We should wait before to merge

**Comment by jouvin** [comment]
@Pansanel in `cluster.build.properties`, could you put `14.10.0` instead of `14.6.0`?

**Comment by jouvin** [comment]
You'll also need to rebase your PR on #21 and move your cluster example into `clusters/openstack`.

**Comment by Pansanel** [comment]
These templates does not compile with 14.10.0 due to https://github.com/quattor/configuration-modules-core/pull/442.

**Comment by jouvin** [comment]
OK. Either we can merge this PR before 15.2 and in this case, it will upgraded by the release script. Or we miss 15.2 and in this case, you could update it to 15.2.0 which should work.
As soon as the reorganisation of cluster examples is merged and you have rebased/moved this PR to the new layout, we can merge it.

**Comment by jrha** [comment]
Please leave this for 15.4.

**Comment by Pansanel** [comment]
This PR requires a PR planned for 15.4: https://github.com/quattor/configuration-modules-core/pull/423
Therefore, this PR should be kept for 15.4.

**Comment by Pansanel** [comment]
@jouvin the example profiles have been moved to clusters/openstack/icehouse. Ready to merge with for 15.4 release.

**Comment by jrha** [comment]
Merge after quattor/configuration-modules-core#423.

File: sites/example/site/databases.pan [hunk 1]
```diff
@@ -39,6 +39,9 @@ final variable DB_IP = nlist(
   escape("hydra.example.org"),"192.168.0.21",
   escape("cream.example.org"),"192.168.0.22",
   escape("argus.example.org"),"192.168.0.23",
+  escape("cloud-controller.example.org"),"30.0.0.2",
```
Comment (stdweird) [inline sites/example/site/databases.pan]: use private IPs in the examples

File: clusters/openstack/icehouse/cluster.build.properties [hunk 2]
```diff
@@ -0,0 +1,2 @@
+cluster.pan.includes=sites/example openstack/icehouse os quattor/14.6.0 standard
```
Comment (stdweird) [inline clusters/openstack/icehouse/cluster.build.properties]: `quattor/14.6.0` ?

File: clusters/openstack/icehouse/profiles/cloud-controller.example.org.pan [hunk 3]
```diff
@@ -0,0 +1,17 @@
+###############################################################################
```
Comment (stdweird) [inline clusters/openstack/icehouse/profiles/cloud-controller.example.org.pan]: remove this header or use pan annotations like http://quattor-pan.readthedocs.org/en/latest/pan-book/pan-book.html#annotations

File: clusters/openstack/icehouse/profiles/cloud-controller.example.org.pan [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+# object template cloud-controller.example.org.pan
+#
+# RESPONSIBLE: Jerome Pansanel
+#
+###############################################################################
+
+
+object template cloud-controller.example.org;
+
+include { 'machine-types/openstack/cloud_controller' };
```
Comment (stdweird) [inline clusters/openstack/icehouse/profiles/cloud-controller.example.org.pan]: remove the curly brackets

File: clusters/openstack/icehouse/profiles/network-node.example.org.pan [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    'bootproto','dhcp',
+    'onboot', 'yes',
+    'type','Ethernet',
+  );
+
+  SELF[PUBLIC_BRIDGE] = nlist(
+    'device',DATA_BRIDGE,
+    'type','OVSBridge',
+    'bootproto','static',
+    'ip','30.0.0.3',
```
Comment (stdweird) [inline clusters/openstack/icehouse/profiles/network-node.example.org.pan]: use `DB_IP`?

File: clusters/openstack/icehouse/repository/config.pan [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+#'/software/repositories' = add_repositories(SITE_REPOSITORY_LIST,YUM_SITE_SNAPSHOT_NS);
+'/software/repositories' = add_repositories(SITE_REPOSITORY_LIST);
+'/software/repositories' = repository_config(SITE_REPOSITORY_CONFIG);
+
+# Repositories related to grid middleware
+# Only if it can be found in the loadpath (else this is a machine not
+# running gLite like a non UI LAL server)
+include { if_exists('repository/config/grid') };
+
+# Add StratusLab repository
```
Comment (stdweird) [inline clusters/openstack/icehouse/repository/config.pan]: really needed?

**Comment by jrha** [comment]
@alvarosimon can you respond to @stdweird's comments?

**Comment by jrha** [comment]
Also consider replacing all nlist with dict and removing the { } around include statements, both are deprecated.

File: clusters/openstack/icehouse/cluster.build.properties (hunk 2 above)
Comment (Pansanel) [inline clusters/openstack/icehouse/cluster.build.properties]: The 14.6.0 (or earlier) has been successfully tested. Newer version of Quattor are not working (issue with ncm-network component). It is fixed in 15.4 release (see https://github.com/quattor/configuration-modules-core/pull/423).

File: sites/example/site/databases.pan (hunk 1 above)
Comment (Pansanel) [inline sites/example/site/databases.pan]: Ok, I will change the code.

File: clusters/openstack/icehouse/profiles/cloud-controller.example.org.pan (hunk 3 above)
Comment (Pansanel) [inline clusters/openstack/icehouse/profiles/cloud-controller.example.org.pan]: Ok, I will change the code.

File: clusters/openstack/icehouse/profiles/cloud-controller.example.org.pan (hunk 4 above)
Comment (Pansanel) [inline clusters/openstack/icehouse/profiles/cloud-controller.example.org.pan]: Ok, I will change the code.

File: clusters/openstack/icehouse/repository/config.pan (hunk 6 above)
Comment (Pansanel) [inline clusters/openstack/icehouse/repository/config.pan]: No, it is not needed. I will fix the code.

**Comment by jrha** [comment]
@Pansanel: any chance of an update?
This might also need rebasing now.

**Comment by Pansanel** [comment]
Nearly to be finished (currently I modify the code to use dict instead of nlist)

**Comment by jrha** [comment]
Okay, thanks!

**Comment by Pansanel** [comment]
Done.

**Comment by jouvin** [comment]
When trying to compile the provided examples, I get the following error:

```
     [panc] validation error [/tmp/scdb-vanilla/cfg/clusters/openstack/icehouse/profiles/compute-node.example.org.pan]
     [panc] user-defined validation failed
     [panc] element path: '/system/network/interfaces/br-eth1/type'
     [panc] element value: OVSBridge
     [panc] type: 'structure_interface' [/tmp/scdb-vanilla/cfg/quattor/dev/components/network/core-schema.pan:108.28-141.1]
     [panc] type: 'structure_network' [/tmp/scdb-vanilla/cfg/quattor/dev/components/network/core-schema.pan:158.26-173.1]
     [panc] type: 'structure_system' [/tmp/scdb-vanilla/cfg/quattor/dev/quattor/types/system.pan:43.25-64.1]
     [panc] type: 'structure_profile' [/tmp/scdb-vanilla/cfg/quattor/dev/quattor/schema.pan:26.26-30.1]
     [panc] path '/' bound to type structure_profile in [/tmp/scdb-vanilla/cfg/quattor/dev/quattor/profile_base.pan:7.12-7.28]
```

Is there a schema extension required and provided by a PR I missed?

**Comment by Pansanel** [comment]
It requires https://github.com/quattor/configuration-modules-core/pull/423 (it is already merged).

**Comment by jouvin** [comment]
Sorry... my fault! I saw that there was this requirement and that it was merged but I forgot that it would be added to template-library-core only at the time of the release... Works fine, merging.

**Pull request merged**